src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala

```diff
@@ -0,0 +1,143 @@
+
+import cats.implicits._
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.headers.OAuth2BearerToken
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.typesafe.config.Config
+import net.ceedubs.ficus.Ficus._
+import org.broadinstitute.dsde.workbench.google.GoogleCredentialModes.Token
+import org.broadinstitute.dsde.workbench.google.{GoogleIamDAO, HttpGoogleIamDAO}
+import org.broadinstitute.dsde.workbench.leonardo.auth.IamProxyAuthProvider.{CacheKey, ProjectAuthCacheKey}
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions.NotebookClusterAction
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions.ProjectAction
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterName
+import org.broadinstitute.dsde.workbench.leonardo.model.{LeoAuthProvider, ServiceAccountProvider}
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, IamPermission}
+
+import scala.concurrent.duration.{FiniteDuration, MINUTES}
+import scala.concurrent.{ExecutionContext, Future}
+
+object IamProxyAuthProvider {
+  private sealed trait CacheKey
+  private case class ProjectAuthCacheKey(userEmail: WorkbenchEmail, userToken: OAuth2BearerToken, googleProject: GoogleProject, executionContext: ExecutionContext) extends CacheKey
+}
+
+class IamProxyAuthProvider(config: Config, serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(config, serviceAccountProvider) {
+
+  // Create implicit actor needed by the GoogleIamDAO.
+  implicit val system = ActorSystem("iam-proxy-auth-actor-system")
+
+  // Load config values.
+  private val cacheEnabled = config.getOrElse("cacheEnabled", true)
+  private val cacheMaxSize = config.getAs[Int]("cacheMaxSize").getOrElse(1000)
+  private val cacheExpiryTime = config.getAs[FiniteDuration]("cacheExpiryTime").getOrElse(FiniteDuration(15, MINUTES))
+  private val applicationName = config.getString("applicationName")
+  private val requiredPermissions: Set[IamPermission] = config.as[Set[String]]("requiredProjectIamPermissions").map(p => IamPermission(p))
+
+  // Cache notebook auth results from the IAM service. This API is called very often by the notebook proxy
+  // and the "list clusters" endpoint.
+  private val notebookAuthCache = CacheBuilder.newBuilder()
+    .expireAfterWrite(cacheExpiryTime.toSeconds, TimeUnit.SECONDS)
+    .maximumSize(cacheMaxSize)
+    .build(
+      new CacheLoader[CacheKey, Future[Boolean]] {
+        def load(key: CacheKey) = {
+          key match {
+            case ProjectAuthCacheKey(userEmail, userToken, googleProject, executionContext) =>
+              checkUserAccessFromIam(userEmail, userToken, googleProject)(executionContext)
+          }
+        }
+      }
+    )
+
+  protected def petGoogleIamDao(token: String)(implicit executionContext: ExecutionContext): GoogleIamDAO = {
```

Comment:
Nit: this is not technically using the pet, but the user's token. So maybe name this `userGoogleIamDao`?

Reply:
Good catch - fixed.